src/dmabuf.rs

```diff
@@ -0,0 +1,435 @@
+        let params = zwp_linux_dmabuf.create_params(qh, GlobalData);
+        Ok(DmabufParams { params })
+    }
+
+    /// Get default dmabuf feedback. Requires version `4`.
+    ///
+    /// On version `3`, use [`DmabufState::modifiers`].
+    pub fn get_default_feedback<D>(
+        &self,
+        qh: &QueueHandle<D>,
+    ) -> Result<zwp_linux_dmabuf_feedback_v1::ZwpLinuxDmabufFeedbackV1, GlobalError>
+    where
+        D: Dispatch<zwp_linux_dmabuf_feedback_v1::ZwpLinuxDmabufFeedbackV1, DmabufFeedbackData>
+            + 'static,
+    {
+        let zwp_linux_dmabuf = self.zwp_linux_dmabuf.with_min_version(4)?;
+        Ok(zwp_linux_dmabuf.get_default_feedback(qh, DmabufFeedbackData::default()))
+    }
+
+    /// Get default dmabuf feedback for given surface. Requires version `4`.
+    ///
+    /// On version `3`, use [`DmabufState::modifiers`].
+    pub fn get_surface_feedback<D>(
+        &self,
+        surface: &wl_surface::WlSurface,
+        qh: &QueueHandle<D>,
+    ) -> Result<zwp_linux_dmabuf_feedback_v1::ZwpLinuxDmabufFeedbackV1, GlobalError>
+    where
+        D: Dispatch<zwp_linux_dmabuf_feedback_v1::ZwpLinuxDmabufFeedbackV1, DmabufFeedbackData>
+            + 'static,
+    {
+        let zwp_linux_dmabuf = self.zwp_linux_dmabuf.with_min_version(4)?;
+        Ok(zwp_linux_dmabuf.get_surface_feedback(surface, qh, DmabufFeedbackData::default()))
+    }
+}
+
+pub trait DmabufHandler: Sized {
+    fn dmabuf_state(&mut self) -> &mut DmabufState;
+
+    /// Server has sent dmabuf feedback information. This may be received multiple
+    /// times by a `ZwpLinuxDmabufFeedbackV1` object.
+    fn dmabuf_feedback(
+        &mut self,
+        conn: &Connection,
+        qh: &QueueHandle<Self>,
+        proxy: &zwp_linux_dmabuf_feedback_v1::ZwpLinuxDmabufFeedbackV1,
+        feedback: DmabufFeedback,
+    );
+
+    /// `wl_buffer` associated with `params` has been created successfully.
+    fn created(
+        &mut self,
+        conn: &Connection,
+        qh: &QueueHandle<Self>,
+        params: &zwp_linux_buffer_params_v1::ZwpLinuxBufferParamsV1,
+        buffer: wl_buffer::WlBuffer,
+    );
+
+    /// Failed to create `wl_buffer` for `params`.
+    fn failed(
+        &mut self,
+        conn: &Connection,
+        qh: &QueueHandle<Self>,
+        params: &zwp_linux_buffer_params_v1::ZwpLinuxBufferParamsV1,
+    );
+
+    /// Compositor has released a `wl_buffer` created through [`DmabufParams`].
+    fn released(&mut self, conn: &Connection, qh: &QueueHandle<Self>, buffer: &wl_buffer::WlBuffer);
+}
+
+/// Builder for a dmabuf backed buffer
+#[derive(Debug)]
+pub struct DmabufParams {
+    params: zwp_linux_buffer_params_v1::ZwpLinuxBufferParamsV1,
+}
+
+impl DmabufParams {
+    /// Add a plane
+    ///
+    /// In version `4`, it is a protocol error if `format`/`modifier` pair wasn't
+    /// advertised as supported.
+    pub fn add(&self, fd: BorrowedFd<'_>, plane_idx: u32, offset: u32, stride: u32, modifier: u64) {
```

Comment:
This brings up a question. Should Smithay's `Dmabuf` type be put into another crate to allow both Smithay and sctk to have a DmabufBuilder and the Weak/Strong dmabuf types?

Although if we do that, Smithay's current use of dmabuf is very graphics api dependent (remember a dmabuf is just a memory object that is given meaning by a format, stride, plane count and modifiers).

Reply:
In that case we could get rid of the `DmabufParams` type and have `create` and `create_immed` methods of `DmabufState ` that take a `Dmabuf`. This could be useful in some cases. Though maybe not most.

These functions would still have to deal with the protocol error if the `format`/`modifier` wasn't advertised.